R/chat-structured.R

```diff
@@ -84,3 +68,33 @@ convert_from_type <- function(x, type) {
     x
   }
 }
+
+list_to_atomic <- function(x, type) {
+  r_type <- switch(
+    type,
+    boolean = "logical",
+    integer = "integer",
+    number = "double",
+    string = "character",
+    cli::cli_abort("Unknown type {type}", .internal = TRUE)
+  )
+
+  if (length(x) == 0) {
+    return(vector(r_type, 0))
+  }
+
+  types <- map_chr(x, typeof)
+  if (r_type == "integer") {
+    x[types == "double"] <- lapply(x[types == "double"], as.integer)
+    types[types == "double"] <- "integer"
+  } else if (r_type == "double") {
+    x[types == "integer"] <- lapply(x[types == "integer"], as.double)
+    types[types == "integer"] <- "double"
+  }
+
+  # also captures NULLs
+  wrong_type <- types != r_type
+  x[wrong_type] <- list(`mode<-`(NA, r_type))
+
+  unlist(x, use.names = FALSE)
```

Comment:
```suggestion
  unlist(x, use.names = FALSE, recursive = FALSE)
```

This may make it take a faster path too, not sure